src/mlpack/methods/ann/sparse_autoencoder_function.hpp

```diff
@@ -0,0 +1,407 @@
+#ifndef __MLPACK_METHODS_ANN_SPARSE_AUTOENCODER_FUNCTION_HPP
+#define __MLPACK_METHODS_ANN_SPARSE_AUTOENCODER_FUNCTION_HPP
```

Comment:
This file looks like it's heavily based on the code in `mlpack/methods/sparse_autoencoder/sparse_autoencoder_function.hpp`; we could do some git wrangling to move that file instead of making a completely new one, and preserve the history that way (it'll take a little bit of work, but shouldn't be that hard).